homeassistant/components/sensor/filter.py

```diff
@@ -0,0 +1,324 @@
+    """Set up the template sensors."""
+    name = config.get(CONF_NAME)
+    entity_id = config.get(CONF_ENTITY_ID)
+    filters = []
+
+    for _filter in config[CONF_FILTERS]:
+        window_size = _filter.get(CONF_FILTER_WINDOW_SIZE)
+        precision = _filter.get(CONF_FILTER_PRECISION)
+
+        if _filter[CONF_FILTER_NAME] == FILTER_NAME_OUTLIER:
+            radius = _filter.get(CONF_FILTER_RADIUS)
+            filters.append(OutlierFilter(window_size=window_size,
+                                         precision=precision,
+                                         entity=entity_id,
+                                         radius=radius))
+        elif _filter[CONF_FILTER_NAME] == FILTER_NAME_LOWPASS:
+            time_constant = _filter.get(CONF_FILTER_TIME_CONSTANT)
+            filters.append(LowPassFilter(window_size=window_size,
+                                         precision=precision,
+                                         entity=entity_id,
+                                         time_constant=time_constant))
+        elif _filter[CONF_FILTER_NAME] == FILTER_NAME_THROTTLE:
+            filters.append(ThrottleFilter(window_size=window_size,
+                                          precision=precision,
+                                          entity=entity_id))
+
+    async_add_devices([SensorFilter(name, entity_id, filters)])
+
+
+class SensorFilter(Entity):
+    """Representation of a Filter Sensor."""
+
+    def __init__(self, name, entity_id, filters):
+        """Initialize the sensor."""
+        self._name = name
+        self._entity = entity_id
+        self._unit_of_measurement = None
+        self._state = None
+        self._filters = filters
+        self._icon = ICON
+
+    @asyncio.coroutine
+    def async_added_to_hass(self):
+        """Register callbacks."""
+        @callback
+        def filter_sensor_state_listener(entity, old_state, new_state):
+            """Handle device state changes."""
+            self._unit_of_measurement = new_state.attributes.get(
+                ATTR_UNIT_OF_MEASUREMENT)
+            self._icon = new_state.attributes.get(ATTR_ICON, ICON)
+
+            self._state = new_state.state
+
+            if self._state == STATE_UNKNOWN:
+                return
+
+            for filt in self._filters:
+                try:
+                    filtered_state = filt.filter_state(self._state)
+                    _LOGGER.debug("%s(%s, %s) -> %s", filt.name,
+                                  self._entity,
+                                  self._state,
+                                  "skip" if filt.skip else filtered_state)
+                    if filt.skip:
+                        return
+                except ValueError:
+                    _LOGGER.warning("Could not convert state: %s to number",
+                                    self._state)
+                finally:
+                    self._state = filtered_state
```

Comment:
Why is this in the finally ? Do you still want this to happen when a `TypeError` (or whatever) is raised? Probably should be moved below the `try…except…finally`.

Reply:
I figured if the filter doesn't work, it could be helpful for a normal user to see the state which blew up the filter. 

Please note that due to chaining this state is not necessarily the state that came from the raw sensor.

Reply:
I think that when an error occurs, we need to fail hard. Otherwise people might start relying on failed state or not even realize that a failed state has occurred.

Reply:
I see your point, do you suggest moving from warning to error and raise the error ?

Reply:
I wouldn't raise the error, as that does nothing but log an exception. Log the error (as that will show up in the issue list) and just make sure that nothing is updated.

To do so, make sure you keep track of the new state in a local variable and only assign to `self._state` after all filters have run through successfully.